src/main/scala/com/fulcrumgenomics/bam/AnnotateMolecularBarcodes.scala

```diff
@@ -0,0 +1,148 @@
+    |with inconsistent records.
+  """,
+  group = ClpGroups.SamOrBam)
+class AnnotateMolecularBarcodes
+( @arg(flag = "i",   doc = "Input BAM file.")                                     val input: PathToBam,
+  @arg(flag = "o",   doc = "Output BAM file.")                                    val output: PathToBam,
+  @arg(flag = "rs1", doc = "The read structure for read 1.")                       val readStructure1: String,
+  @arg(flag = "rs2", doc = "The read structure for read 2.")                       val readStructure2: String,
+  @arg(flag = "b1",  doc = "SAM tags in which to store the molecular barcodes (one-per segment) for read.")
+                                                                                  val molecularBarcodeTags1: Seq[String] = Seq.empty,
+  @arg(flag = "b2",  doc = "SAM tags in which to store the molecular barcodes (one-per segment).")
+                                                                                  val molecularBarcodeTags2: Seq[String] = Seq.empty,
+  @arg(flag = "a",   doc = "Annotate the read names with the molecular barcodes.") val annotateReadNames: Boolean = true,
+  @arg(flag = "t",   doc = "Annotate both reads from a read pair with each others tags.")
+                                                                                  val tagBothEnds: Boolean = true
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+  molecularBarcodeTags1.foreach(tag => if (tag.length != 2) throw new ValidationException("SAM tags must be of length two: " + tag))
+  molecularBarcodeTags2.foreach(tag => if (tag.length != 2) throw new ValidationException("SAM tags must be of length two: " + tag))
+  if (readStructure1.isEmpty) throw new ValidationException("readStructure1 may not be empty")
+  if (readStructure2.isEmpty) throw new ValidationException("readStructure2 may not be empty")
+  if (tagBothEnds) {
+    molecularBarcodeTags1.foreach { tag => assert(!molecularBarcodeTags2.contains(tag)) }
+    molecularBarcodeTags2.foreach { tag => assert(!molecularBarcodeTags1.contains(tag)) }
+  }
+
+  override def execute(): Unit = {
+    val rs1 = ReadStructure(readStructure1)
+    val rs2 = ReadStructure(readStructure2)
+    assert(molecularBarcodeTags1.length == rs1.count {
+      case m: MolecularBarcode => true
+      case _ => false
+    })
+    assert(molecularBarcodeTags2.length == rs2.count {
+      case m: MolecularBarcode => true
+      case _ => false
+    })
+    val in: SamReader = SamReaderFactory.make.open(input.toFile)
+    val iterator: SAMRecordIterator = in.iterator()
+    val out: SAMFileWriter = new SAMFileWriterFactory().makeSAMOrBAMWriter(in.getFileHeader, true, output.toFile)
+    iterator.sliding(2,2).foreach {
+      case Seq(r1: SAMRecord, r2: SAMRecord) =>
+        assert(r1.getReadName.compareTo(r2.getReadName) == 0)
```

Comment:
I think also it should assert that all the reads coming through are unmapped, since this will mangle things horribly if they are mapped and the tool states that mapped reads are not supported.